src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs

```diff
@@ -358,28 +358,41 @@ private void WriteAsPaxSharedInternal(Span<byte> buffer, out long actualLength)
             _checksum = WriteChecksum(tmpChecksum, buffer);
         }
 
-        // All formats save in the name byte array only the ASCII bytes that fit.
+        // All formats save in the name byte array only the UTF8 bytes that fit.
         private int WriteName(Span<byte> buffer)
         {
-            ReadOnlySpan<char> src = _name.AsSpan(0, Math.Min(_name.Length, FieldLengths.Name));
-            Span<byte> dest = buffer.Slice(FieldLocations.Name, FieldLengths.Name);
-            int encoded = Encoding.ASCII.GetBytes(src, dest);
-            return Checksum(dest.Slice(0, encoded));
+            ReadOnlySpan<char> name = _name;
+            int utf16NameTruncatedLength = GetUtf16TruncatedTextLength(name, FieldLengths.Name);
+
+            ReadOnlySpan<char> truncatedName = name.Slice(0, utf16NameTruncatedLength);
+            Span<byte> destination = buffer.Slice(FieldLocations.Name, FieldLengths.Name);
+            Encoding.UTF8.GetBytes(truncatedName, destination);
+
+            return Checksum(destination);
         }
 
-        // Ustar and PAX save in the name byte array only the ASCII bytes that fit, and the rest of that string is saved in the prefix field.
+        // Ustar and PAX save in the name byte array only the UTF8 bytes that fit, and the rest of that string is saved in the prefix field.
         private int WritePosixName(Span<byte> buffer)
         {
-            int checksum = WriteName(buffer);
+            ReadOnlySpan<char> name = _name;
+            int utf16NameTruncatedLength = GetUtf16TruncatedTextLength(name, FieldLengths.Name);
 
-            if (_name.Length > FieldLengths.Name)
+            ReadOnlySpan<char> truncatedName = name.Slice(0, utf16NameTruncatedLength);
+            Span<byte> destination = buffer.Slice(FieldLocations.Name, FieldLengths.Name);
+            Encoding.UTF8.GetBytes(truncatedName, destination);
+
+            int checksum = Checksum(destination);
+
+            if (utf16NameTruncatedLength < name.Length)
```

Comment:
What's the expected behavior if truncatedName as UTF8 only takes up, say, 98 of the 100 bytes in the Name field and then we spill over to Prefix for the rest?  How do tools know the extra two bytes in Name aren't actually part of the name?

Reply:
the thing is, the prefix logic is wrong, I filed https://github.com/dotnet/runtime/issues/75360 for it. I don't want to fix it here because of time.

Reply:
It's not clear to me from that issue: is the problem with how we read/interpret name and prefix, or is the issue with how we write out the name and prefix, or both?

Reply:
the issue is how we write it.
We currently do [this](https://github.com/dotnet/runtime/blob/55060ebf68e16f2838d30060d41cd71b0bc9c540/src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs#L370): 
> Ustar and PAX save in the name byte array only the bytes that fit, and the rest of that string is saved in the prefix field. 

Which is actually a suffix.

Instead we should do [this](https://www.freebsd.org/cgi/man.cgi?tar(5)):
> If the pathname is too long to fit in the 100 bytes provided by the standard format, it can be split **at any / character** with the first portion going into the prefix field.  If the prefix field is not empty, the reader will prepend the prefix value and a / character to the regular name field to obtain the full pathname. The standard does not require a trailing / character on directory names, though most implementations still include this for compatibility reasons.

Reply:
Then if the code that's there is 100% wrong and the code in the PR it's being updated to is 100% wrong, why are we changing that code?  Seems like we should either fix it in this PR, or as part of changing it in this PR, this should fail in a more reliable way rather than outputting corrupt data.  Do we plan to fix this issue for 7.0 as well?